### PR TITLE
Log and config file generation as 0640 root:adm

### PR DESCRIPTION
**Pull request by dbungert**
* Implement file mode 0640, owner root:adm for log files and config files that will be copied to the target system and may contain sensitive information.
* Centralize on subiquity-specific write_file & helper funtions rather than a mix of curtin / cloud-init / subiquity versions

### CODE REVIEW

**Review by ogayot — COMMENTED** [review]
I don't have much context to lean on. But I think we want to keep more restrictive permissions for `/autoinstall.yaml`

File: subiquity/server/server.py [hunk 1]
```diff
@@ -516,11 +516,8 @@ async def wait_for_cloudinit(self):
                 autoinstall_path = '/autoinstall.yaml'
                 if 'autoinstall' in self.cloud.cfg:
                     if not os.path.exists(autoinstall_path):
-                        atomic_helper.write_file(
-                            autoinstall_path,
-                            safeyaml.dumps(
-                                self.cloud.cfg['autoinstall']).encode('utf-8'),
-                            mode=0o600)
+                        cfg = self.cloud.cfg['autoinstall']
+                        write_file(autoinstall_path, safeyaml.dumps(cfg))
```
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/server/server.py:520]: Intentionally going to 0640

File: subiquitycore/log.py [hunk 2]
```diff
@@ -17,12 +17,14 @@
 import logging
 import os
 
+from subiquitycore.file_util import _DEF_PERMS, _DEF_GROUP
+
 
 def setup_logger(dir, base='subiquity'):
     os.makedirs(dir, exist_ok=True)
     if os.getuid() == 0:
         os.chmod(dir, 0o775)
```
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquitycore/log.py:26]: Good point!
0750 would be consistent with the 0640 on the files.

File: subiquity/server/server.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/server/server.py:529]: It's worth noting that this particular file doesn't get copied to the target environment.  So I see no reason to keep a chmod override.

File: subiquitycore/file_util.py [hunk 3]
```diff
@@ -1,32 +1,67 @@
...
 
 
-def write_file(filename, content, mode=None, omode="wb", copy_mode=False):
-    """Atomically write filename.
-    open filename in mode 'omode', write content, chmod to 'mode'.
-    """
-    if mode is None:
-        mode = _DEF_PERMS
+@contextlib.contextmanager
+def open_perms(filename, *, cmode=None, omode='w', copy_mode=False):
```
Comment (mwhudson) [inline subiquitycore/file_util.py:28]: No caller passes omode or copy_mode that I can see. Can we imagine a situation where we might use them?

File: subiquitycore/file_util.py [hunk 4]
```diff
@@ -1,32 +1,67 @@
...
+
+def write_file(filename, content, **kwargs):
+    """Atomically write filename.
+    open filename in mode 'omode', write content, chmod to 'cmode'.
+    """
+    with open_perms(filename, **kwargs) as tf:
+        tf.write(content)
+
+
+def generate_config(filename, content, **kwargs):
```
Comment (mwhudson) [inline subiquitycore/file_util.py:63]: Could this function do the yaml.dump part too?

**Review by mwhudson — APPROVED** [review]
Just a couple of comments. Sorry for the slow review!

File: subiquitycore/file_util.py (hunk 3 above)
Comment (dbungert) [reply subiquitycore/file_util.py:28]: omode used to be used but disappears in this branch.  copy_mode was never used.  I'll remove both.

**Pull request merged**